fdbclient/NativeAPI.actor.cpp

```diff
@@ -5919,8 +6021,42 @@ ACTOR Future<Version> extractReadVersion(Reference<TransactionState> trState,
 	return rep.version;
 }
 
+bool rkThrottlingCooledDown(DatabaseContext* cx, TransactionPriority priority) {
+	if (priority == TransactionPriority::IMMEDIATE) {
+		return true;
+	} else if (priority == TransactionPriority::BATCH) {
+		if (cx->lastRkBatchThrottleTime == 0.0) {
+			return true;
+		}
+		return (now() - cx->lastRkBatchThrottleTime > CLIENT_KNOBS->GRV_CACHE_RK_COOLDOWN);
+	} else if (priority == TransactionPriority::DEFAULT) {
+		if (cx->lastRkDefaultThrottleTime == 0.0) {
+			return true;
+		}
+		return (now() - cx->lastRkDefaultThrottleTime > CLIENT_KNOBS->GRV_CACHE_RK_COOLDOWN);
+	}
+	return false;
+}
+
 Future<Version> Transaction::getReadVersion(uint32_t flags) {
 	if (!readVersion.isValid()) {
+		if (!CLIENT_KNOBS->FORCE_GRV_CACHE_OFF && !trState->options.skipGrvCache &&
+		    (deterministicRandom()->random01() <= CLIENT_KNOBS->DEBUG_USE_GRV_CACHE_CHANCE ||
+		     trState->options.useGrvCache) &&
+		    rkThrottlingCooledDown(getDatabase().getPtr(), trState->options.priority)) {
+			// Upon our first request to use cached RVs, start the background updater
+			if (!trState->cx->grvUpdateHandler.isValid()) {
+				trState->cx->grvUpdateHandler = backgroundGrvUpdater(getDatabase().getPtr());
+			}
+			Version rv = trState->cx->getCachedReadVersion();
+			double lastTime = trState->cx->getLastGrvTime();
+			double requestTime = now();
+			if (requestTime - lastTime <= CLIENT_KNOBS->MAX_VERSION_CACHE_LAG && rv != Version(0)) {
+				ASSERT(!debug_checkVersionTime(rv, requestTime, "CheckStaleness"));
+				readVersion = rv;
+				return readVersion;
+			} // else go through regular GRV path
```

Comment:
One other thing that we might want is a transaction option for "minimum read version" that gets verified here. The idea there is that it can be used as a sort of "causality" token, so that if a client talks to one server A which in turn commits to FDB, it can then forward the commit version from A as a causality token to a different server B so that server B is guaranteed to see server A's commit. You can see that field here on the `WeakReadSemantics` object: https://github.com/FoundationDB/fdb-record-layer/blob/5efeb2355be9507a9802d18439e53016dfc1cac3/fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBDatabase.java#L1314

This could also be used in the single instance case in the presence of the multi-threaded or multi-version client. For example, if you had two threads talking to the same FDB database, with client transactions being randomly assigned to each thread, you could make sure that each client sees its own writes by forwarding along the commit version as a "minimum read version" so even if it winds up on the other thread, it still sees its own commits.

Reply:
I think both this and the other proposal for bounding the staleness per-transaction are interesting ideas. What do you think about raising them as GitHub issues to be added as future enhancements?

Reply:
Sure thing. Filed:

* Cached read versions should allow staleness bounds to be specified as a transaction option #6420
* Cached read versions should have a minimum version transaction option to guarantee causality #6421